### PR TITLE
Fix typo: `principle` to `principal`

### LINKED ISSUE #110: Typo: `principle` => `principal`
## Description

As described in this issue: https://github.com/ExocoreNetwork/exocore-contracts/issues/33, we should use `principle` to represent user's collaterals while we incorrectly spell it as `principle`

### PR DESCRIPTION
**Pull request by adu-web3**
## Description

----

Closes #110 

I have searched all `principle`s that appear in code, doc or other files and replace them with `principal` if it means staked collaterals. Also the abi json file of `IAssets.sol` is re-generated.

This includes commits from PR #92 and could be merged after PR #92

## Summary by CodeRabbit

- **New Features**
  - Introduced new functions for depositing and withdrawing assets, as well as querying client chain indices.
  - Added a new interface and related functions to interact with the assets module directly.

- **Refactor**
  - Renamed packages and variables for consistency across the codebase.
  - Modified method names and parameters to streamline asset operations.
  - Combined deposit and withdraw functionalities into a single method for better efficiency.

- **Tests**
  - Added comprehensive test cases for deposit, withdrawal, and client chain retrieval functionalities.
  - Introduced a new test suite for precompiling assets to ensure robust testing.

- **Bug Fixes**
  - Corrected a typo in a constant name for better code readability and accuracy.

### CODE REVIEW

File: precompiles/assets/IAssets.sol [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+/// TRANSACTIONS
+/// @dev deposit the client chain assets for the staker,
+/// that will change the state in deposit module
+/// Note that this address cannot be a module account.
+/// @param clientChainLzID The LzID of client chain
+/// @param assetsAddress The client chain asset address
+/// @param stakerAddress The staker address
+/// @param opAmount The amount to deposit
+    function depositTo(
+        uint32 clientChainLzID,
```
Comment (bwhour) [inline precompiles/assets/IAssets.sol]: to be compatible with other non-layerzero clientchains. it's better to rename this clientChainLzID to clientChainID

File: precompiles/assets/IAssets.sol [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+/// TRANSACTIONS
+/// @dev withdraw To the staker, that will change the state in withdraw module
+/// Note that this address cannot be a module account.
+/// @param clientChainLzID The LzID of client chain
+/// @param assetsAddress The client chain asset Address
+/// @param withdrawAddress The withdraw address
+/// @param opAmount The withdraw amount
+    function withdrawPrincipal(
+        uint32 clientChainLzID,
```
Comment (bwhour) [inline precompiles/assets/IAssets.sol:39]: to be compatible with other non-layerzero clientchains. it's better to rename this clientChainLzID to clientChainID

File: precompiles/assets/tx.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+	// DepositAndWithdraw transaction.
+	MethodDepositTo = "depositTo"
+	MethodWithdraw  = "withdrawPrincipal"
+
+	MethodGetClientChains = "getClientChains"
+)
+
+// DepositAndWithdraw deposit and withdraw the client chain assets for the staker,
+// that will change the state in assets module.
+func (p Precompile) DepositAndWithdraw(
```
Comment (bwhour) [inline precompiles/assets/tx.go]: Unify all "DepositAndWithdraw" instances into "DepositOrWithdraw."

File: precompiles/assets/types.go [hunk 4]
```diff
@@ -1,29 +1,29 @@
...
+	assetskeeper "github.com/ExocoreNetwork/exocore/x/assets/keeper"
 	"github.com/ExocoreNetwork/exocore/x/assets/types"
-	"github.com/ExocoreNetwork/exocore/x/deposit/keeper"
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	cmn "github.com/evmos/evmos/v14/precompiles/common"
 	"golang.org/x/xerrors"
 )
 
-func (p Precompile) GetDepositToParamsFromInputs(ctx sdk.Context, args []interface{}) (*keeper.DepositParams, error) {
+func (p Precompile) GetDepositWithdrawParamsFromInputs(ctx sdk.Context, args []interface{}) (*assetskeeper.DepositWithdrawParams, error) {
```
Comment (bwhour) [inline precompiles/assets/types.go]: depositOrWithdrawParams /  DepositOrWithdrawParams will be better.

File: precompiles/assets/assets.go [hunk 5]
```diff
@@ -89,12 +86,26 @@ func (p Precompile) Run(evm *vm.EVM, contract *vm.Contract, readOnly bool) (bz [
 	// It avoids panics and returns the out of gas error so the EVM can continue gracefully.
 	defer cmn.HandleGasError(ctx, contract, initialGas, &err)()
 
-	if method.Name == MethodDepositTo {
-		bz, err = p.DepositTo(ctx, evm.Origin, contract, stateDB, method, args)
+	switch method.Name {
+	// transactions
+	case MethodDepositTo, MethodWithdraw:
+		bz, err = p.DepositAndWithdraw(ctx, evm.Origin, contract, stateDB, method, args)
```
Comment (bwhour) [inline precompiles/assets/assets.go]: Unify all "DepositAndWithdraw" instances into "DepositOrWithdraw."

File: precompiles/assets/assets_test.go [hunk 6]
```diff
@@ -0,0 +1,418 @@
...
+	}
+}
+
+func paddingClientChainAddress(input []byte, outputLength int) []byte {
+	if len(input) < outputLength {
+		padding := make([]byte, outputLength-len(input))
+		return append(input, padding...)
+	}
+	return input
+}
```
[the remark this answers is not in the document]
Comment (adu-web3) [reply precompiles/assets/assets_test.go]: please help create an issue to fix it in future pr

File: precompiles/assets/abi.json [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+      },
+      {
+        "internalType": "uint256",
+        "name": "latestAssetState",
+        "type": "uint256"
+      }
+    ],
+    "stateMutability": "nonpayable",
+    "type": "function"
+  }
```
[the remark this answers is not in the document]
Comment (adu-web3) [reply precompiles/assets/abi.json]: fixed

File: precompiles/assets/IAssets.sol (hunk 1 above)
Comment (adu-web3) [reply precompiles/assets/IAssets.sol]: @TimmyExogenous said that he would address this in a new pr, so I would resolve these conversations. Same for below conversations

File: precompiles/assets/tx.go [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+		return method.Outputs.Pack(false, nil)
+	}
+	ids := make([]uint32, 0, len(infos))
+	for id := range infos {
+		// #nosec G701 // already checked
+		convID := uint32(id)
+		ids = append(ids, convID)
+	}
+	return method.Outputs.Pack(true, ids)
+}
```
[the remark this answers is not in the document]
Comment (adu-web3) [reply precompiles/assets/tx.go]: please help create an issue and it would be addressed in a new pr

**Review by TimmyExogenous — APPROVED** [review]

**Review by MaxMustermann2 — APPROVED** [review]

**Review by cloud8little — APPROVED** [review]

**Pull request merged**